File: war3map/Generic/Point.ts
export class Point {
    public x: number;
    public y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    public distanceTo(target: Point): number {
        return (math.sqrt(this.x - target.x) + math.sqrt(this.y - target.y));
    }

    public directionTo(target: Point) {
        let radians = math.atan(target.y - this.y, target.x - this.x);
        return (radians * 180 / Math.PI);
    }

    public toLocation() {
        return Location(this.x, this.y);
    }

    public polarProject(distance: number, angle: number): Point {
        let x = this.x + distance * math.cos(angle * bj_DEGTORAD);
        let y = this.y + distance * math.sin(angle * bj_DEGTORAD);
        return new Point(x, y);
    }

    public static fromLocation(inputLoc: location) {
        return new Point(GetLocationX(inputLoc), GetLocationY(inputLoc));
    }

    public static fromLocationClean(inputLoc: location) {
        let point = new Point(GetLocationX(inputLoc), GetLocationY(inputLoc));
        RemoveLocation(inputLoc);
        return point;
    }

    distanceToLine(lineStart: Point, lineEnd: Point) {
        let A = this.x - lineStart.x;
        let B = this.y - lineStart.y;
        let C = lineEnd.x - lineStart.x;
        let D = lineEnd.y - lineStart.y;

        let dot = A * C + B * D;
        let len_sq = C * C + D * D;
        let param = -1;
        if (len_sq != 0) //in case of 0 length line
            param = dot / len_sq;

        let xx, yy;

        if (param < 0) {
            xx = lineStart.x;
            yy = lineStart.y;
        } else if (param > 1) {
            xx = lineEnd.x;
            yy = lineEnd.y;
        } else {
            xx = lineStart.x + param * C;
            yy = lineStart.y + param * D;
        }

        let dx = this.x - xx;
        let dy = this.y - yy;
        return Math.sqrt(dx * dx + dy * dy);
    }
}

File: src/war3map/Generic/Point.lua
local ____exports = {}
____exports.Point = {}
local Point = ____exports.Point
Point.name = "Point"
Point.__index = Point
Point.prototype = {}
Point.prototype.__index = Point.prototype
Point.prototype.constructor = Point
function Point.new(...)
    local self = setmetatable({}, Point.prototype)
    self:____constructor(...)
    return self
end
function Point.prototype.____constructor(self, x, y)
    self.__loc = nil
    self.x = x
    self.y = y
end
function Point.prototype.distanceTo(self, target)
    return math.sqrt(((self.x - target.x) * (self.x - target.x)) + ((self.y - target.y) * (self.y - target.y)))
end
function Point.prototype.directionTo(self, target)
    local radians = math.atan(target.y - self.y, target.x - self.x)
    return (radians * 180 / math.pi)
end
function Point.prototype.cleanseLastLocation(self)
    if self.__loc ~= nil then
        RemoveLocation(self.__loc)
    end
end
function Point.prototype.toLocation(self)
    self.__loc = Location(self.x, self.y)
    return self.__loc
end
function Point.prototype.updateToLocation(self, inputLoc)
    self.x = GetLocationX(inputLoc)
    self.y = GetLocationY(inputLoc)
end
function Point.prototype.updateToLocationClean(self, inputLoc)
    self.x = GetLocationX(inputLoc)
    self.y = GetLocationY(inputLoc)
    RemoveLocation(inputLoc)
end
function Point.prototype.polarProject(self, distance, angle)
    local x = self.x + distance * math.cos(angle * bj_DEGTORAD)
    local y = self.y + distance * math.sin(angle * bj_DEGTORAD)
    return ____exports.Point.new(x, y)
end
function Point.fromLocation(self, inputLoc)
    return ____exports.Point.new(
        GetLocationX(inputLoc),
        GetLocationY(inputLoc)
    )
end
function Point.fromLocationClean(self, inputLoc)
    local point = ____exports.Point.new(
        GetLocationX(inputLoc),
        GetLocationY(inputLoc)
    )
    RemoveLocation(inputLoc)
    return point
end
function Point.fromUnit(self, inputU)
    local loc = GetUnitLoc(inputU)
    local point = ____exports.Point.new(
        GetLocationX(loc),
        GetLocationY(loc)
    )
    RemoveLocation(loc)
    return point
end
function Point.prototype.distanceToLine(self, lineStart, lineEnd)
    local A = self.x - lineStart.x
    local B = self.y - lineStart.y
    local C = lineEnd.x - lineStart.x
    local D = lineEnd.y - lineStart.y
    local dot = A * C + B * D
    local len_sq = C * C + D * D
    local param = -1
    if len_sq ~= 0 then
        param = dot / len_sq
    end
    local xx
    local yy
    if param < 0 then
        xx = lineStart.x
        yy = lineStart.y
    elseif param > 1 then
        xx = lineEnd.x
        yy = lineEnd.y
    else
        xx = lineStart.x + param * C
        yy = lineStart.y + param * D
    end
    local dx = self.x - xx
    local dy = self.y - yy
    return math.sqrt(dx * dx + dy * dy)
end
function Point.prototype.__tostring(self)
    return "point {x:" .. tostring(self.x) .. ", y:" .. tostring(self.y) .. " }"
end
return ____exports
